source/material_model/cpp.cc

```diff
@@ -0,0 +1,623 @@
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/cpp.h>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    template <int dim>
+    bool
+    CPP<dim>::
+    is_compressible () const
+    {
+      return is_compressible_param;
+    }
+
+    template <int dim>
+    double
+    CPP<dim>::
+    reference_viscosity () const
+    {
+      return reference_viscosity_param;
+    }
+
+    template <int dim>
+    void
+    CPP<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      user_eval (in, out, this);
+    }
+
+    /*
+     * Attempt to locate an indenter command to beautify the generated
+     * C++ code.
+     *
+     * @return Command line tool for indenting c++ source code, in a format
+     * such that its only missing argument is the file name, to be inserted
+     * at the end of the string.
+     */
+    std::string
+    find_indenter()
+    {
+      if (execute("which astyle >/dev/null") == 0)
+        if (execute("[ -f '" ASPECT_SOURCE_DIR "/doc/astyle.rc' ]") == 0)
+          return "astyle --options=" ASPECT_SOURCE_DIR "/doc/astyle.rc";
+        else
+          return "astyle";
+      else if (execute("which astyle2.04 >/dev/null") == 0)
+        if (execute("[ -f '" ASPECT_SOURCE_DIR "/doc/astyle.rc' ]") == 0)
+          return "astyle2.04 --options=" ASPECT_SOURCE_DIR "/doc/astyle.rc";
+        else
+          return "astyle2.04";
+      else if (execute("which indent >/dev/null") == 0)
+        return "indent";
+      else if (execute("which clang-format >/dev/null") == 0)
+        return "clang-format";
+
+      return "";
+    }
+
+    /*
+     * Make sure user didn't specify 'none' along with other nonlinear
+     * dependencies.
+     */
+    void
+    assert_valid_deps (std::vector<std::string> list, std::string param_name)
+    {
+      AssertThrow(list.size() == 1
+                  || (std::find(list.begin(),list.end(),"none") == list.end()),
+                  ExcMessage("The property 'none' in the parameter '"+param_name+"' "
+                             "cannot be combined with any other options."));
+    }
+
+    /*
+     * Write a simple CMakeLists.txt for the generated code, to simplify
+     * linking with Aspect and Deal.II. Basically, treat the code as a plugin,
+     * since it occupies the same namespace, and should expose most of the same
+     * interfaces as a native plugin.
+     */
+    void
+    generate_cmakelists (const std::string cdir)
```

Comment:
same here and for the other functions below: non-scalar input arguments should be const reference